sdk/core/Azure.Core/src/Shared/EventSourceEventFormatting.cs

```diff
@@ -1,79 +1,125 @@
+        if (!string.IsNullOrWhiteSpace(eventData.Message))
+        {
+            stringBuilder.AppendLine();
+            stringBuilder.Append(nameof(eventData.Message)).Append(" = ").Append(eventData.Message);
         }
 
-        private static void ProcessPayloadArray(object?[] payloadArray)
+        if (eventData.PayloadNames != null)
         {
-            for (int i = 0; i < payloadArray.Length; i++)
+            for (int i = 0; i < eventData.PayloadNames.Count; i++)
             {
-                payloadArray[i] = FormatValue(payloadArray[i]);
+                stringBuilder.AppendLine();
+                stringBuilder.Append(eventData.PayloadNames[i]).Append(" = ").Append(payloadArray[i]);
             }
         }
 
-        private static object? FormatValue(object? o)
+        return ToStringAndReturnStringBuilder(stringBuilder);
+    }
+
+    private static void ProcessPayloadArray(object?[] payloadArray)
+    {
+        for (int i = 0; i < payloadArray.Length; i++)
         {
-            if (o is byte[] bytes)
+            payloadArray[i] = FormatValue(payloadArray[i]);
+        }
+    }
+
+    private static object? FormatValue(object? o)
+    {
+        if (o is byte[] bytes)
+        {
+#if NET6_0_OR_GREATER
+            return Convert.ToHexString(bytes);
+#else
+            int length = 2 * bytes.Length; // Two hex characters per byte
+            if (length < 256)
             {
-                var stringBuilder = new StringBuilder();
-                foreach (byte b in bytes)
+                static char ToHex(int value) => (char)(value < 10 ? value + '0' : value - 10 + 'A');
+
+                Span<char> buffer = stackalloc char[length];
+                for (int i = 0; i < bytes.Length; i++)
                 {
-                    stringBuilder.AppendFormat(CultureInfo.InvariantCulture, "{0:X2}", b);
+                    byte b = bytes[i];
+                    buffer[i * 2] = ToHex(b >> 4);
+                    buffer[i * 2 + 1] = ToHex(b & 0xF);
                 }
 
-                return stringBuilder.ToString();
+                return buffer.ToString();
             }
 
-            return o;
+            var stringBuilder = new StringBuilder(length);
+            foreach (byte b in bytes)
+            {
+                stringBuilder.AppendFormat(CultureInfo.InvariantCulture, "{0:X2}", b);
+            }
+
+            return stringBuilder.ToString();
+#endif
+        }
+
+        return o;
+    }
+
+    private static StringBuilder RentStringBuilder()
+    {
+        StringBuilder? builder = s_cachedStringBuilder;
+        if (builder is null)
+        {
+            return new StringBuilder(CachedStringBuilderCapacity);
         }
+
+        s_cachedStringBuilder = null;
+        return builder;
+    }
+
+    private static string ToStringAndReturnStringBuilder(StringBuilder builder)
+    {
+        string result = builder.ToString();
+        if (builder.Capacity <= CachedStringBuilderCapacity)
```

Comment:
If `s_cachedStringBuilder` is not `null` here - do we want to clear and update the reference or just drop this new instance on the floor?

Reply:
If the Rent/Return pattern is followed correctly, and there are no nested calls, then `s_cachedStringBuilder` should always be `null` at this point. The use of `[ThreadStatic]` also isolates modifications from concurrent threads.
And, if those design invariants fail, then there's no harm overwriting s_cachedStringBuilder. It doesn't break.

Reply:
Agreed. I couldn't see a case with the thread local use that would cause a race, but rending seems to account for the "someone already rented this" case (though probably intended more for the initial instantiation) so I felt it worth mentioning.